### DISCUSSION TITLE
Is there a way to use decorator with setters?

### DISCUSSION THREAD

**Discussion by UserAbuser228** [mobxjs/mobx#2820] [Q&A]
I would like to use decorator, that would control property's setter. Here's what I tried.

```typescript
const MyDecorator = <T>(): PropertyDecorator => () => {
    let value: T;
    return {
        configurable: true,
        get: () => value,
        set: val => {
            value = val;
        },
    };
};

class MyClass {
    @MyDecorator()
    field = 1;

    constructor() {
        makeAutoObservable(this);
        reaction(() => this.field, () => console.log('reacted'));
    }
}

const myClass = new MyClass();
console.log(myClass.field++);
console.log(myClass.field++);
console.log(myClass.field++);
```

In the case above, both my `@MyDecorator` and `makeAutoObservable` don't work. They do work separatly. Without `makeAutoObservable` the value is increasing as it should be. Without `@MyDecorator` it's increasing and runs the callback for `reaction`. Both of them do not actually change the value, it always equal to 1. 

I tried several ways, even with older versions and `@observable` decorators, but none of my tries were successfull. 

So, could you please tell is there a way to create a decorator with setter?

**Comment by urugator** [comment]
The getter(+setter) fields are transformed to `computed`. You need the underlying value being observable. No promises, but you can try:
```javascript
const MyDecorator = <T>(): PropertyDecorator => () => {
    const value = observable.box();
    return {
        configurable: true,
        get: () => value.get(),
        set: val => value.set(val)
    };
};
```

**Reply by UserAbuser228** [reply]
Thank you, this code can help me. But there are some stange things. If I do `reaction(() => this.field, ...)` everything is OK, but if I use `observe(this, ...)` the `field` is always undefined

**Reply by urugator** [reply]
Do you mean `observe(this, listener)` or `observe(this, "field", listener)`?
The first one observes changes of object, but `field` is just a function (getter) that does not change over time.
The latter basically creates the `reaction` which is a bit measleding, `observe` on computed value is conceptually a bit wierd (it shouldn't be supported imo, but whatever...)

**Reply by UserAbuser228** [reply]
Oh, sorry, I meant `observe(this, listener)`. Okay, but because of adding `observe(this, listener)` I can't read `field` at all, not only in the `listener`

**Reply by urugator** [reply]
Can you post complete example?

**Reply by UserAbuser228** [reply]
Sure. It's the 5th version

```typescript
const MyDecorator = (): PropertyDecorator => () => {
    const value = observable.box();
    return {
        configurable: true,
        get: () => value.get(),
        set: val => {
            value.set(val);
        },
    };
};

export class MyClass {
    @MyDecorator()
    @observable field = [1];

    constructor() {
        observe(this, v => console.log(`reacted ${v}`));
    }
}
const myClass = new MyClass();
console.log(myClass.field); // undefined;
```

**Reply by urugator** [reply]
Firstly you have to call `makeObservable` from `constructor` (the alone `@observable` does nothing), then it should throw that you cannot apply `@observable` to getter/setter property (not sure what is the decorator application order here, `MyDecorator` must be called first). Generally speaking we don't support combining decorators, but applying custom decorator before our decorator can work. Can't help more, because I am not sure what you're trying to do...

**Reply by UserAbuser228** [reply]
All I wanted is to make a decorator, that could set initial value for the observable field and later react on that value change. Something like
```typescript
class Class {
  @observable field = someFunc('token') ?? initialValue;

  constructor () {
    reaction(() => this.field, () => someOtherFunc('token'));
  }
}
```

But replaced with

```typescript
class Class {
  @Decorator('token')
  @observable field = initialValue;
}
```

Also, using `@computed` instead of `@observable` throws an exception 
```
Uncaught Error: [mobx] Trying to declare a computed value for unspecified getter 'field'
```

**Reply by urugator** [reply]
Ehmm wait a sec ... property decorators cannot redefine property - you can't return descriptor from property decorator...

**Reply by UserAbuser228** [reply]
That's not true. The decorator does work, if the property is not observable

**Reply by urugator** [reply]
That's wierd
https://www.typescriptlang.org/docs/handbook/decorators.html#property-decorators:
*A Property Descriptor is not provided as an argument to a property decorator due to how property decorators are initialized in TypeScript. This is because there is currently no mechanism to describe an instance property when defining members of a prototype, and no way to observe or modify the initializer for a property. **The return value is ignored too.** As such, a property decorator can only be used to observe that a property of a specific name has been declared for a class.*

**Reply by urugator** [reply]
Anyway why do you need getter/setter, can't you just return `{ value: someFunc(token) }` from decorator?

**Reply by UserAbuser228** [reply]
Because that will make the field readonly

**Reply by urugator** [reply]
Ok, I thought it's obvious, but: `{ value: someFunc(token), writable: true, enumerable: true, configurable: true }`.
Or do you mean that you want to make the field readonly?

**Reply by UserAbuser228** [reply]
Excuse me, not obvious for me. That is still not what I need. This way I can set initial value, but what about reacting on the value's change?
I just wander can my idea of decorator
```typescript
@Decorator('token')
@observable field = initialValue; // works only if (someFunc('token') === null);
```
can exists.

**Reply by urugator** [reply]
> but what about reacting on the value's change?

Just make the field observable as usual: `@observable` + `makeObservable(this)` or `makeObservable(this, { field: observable })`. `makeObservable` should use the updated descriptor instead of the original one.
You can check if the decorator was applied correctly - in constructor, before calling `makeObservable` -  `console.log(Object.getOwnPropertyDescriptor(this, "field"))` - you should see the descriptor returned from your decorator.

**Comment by mweststrate** [comment]
Depends on how decorators are transpiled, the semantics of Babel are close
but different to TS

**Reply by UserAbuser228** [reply]
Could you please tell me if such a [decorator](https://github.com/mobxjs/mobx/discussions/2820#discussioncomment-390025) is even possible?

**Reply by mweststrate** [reply]
I strongly recommend against writing your own decorators. There are currently 3 different proposals / implementations you will have to deal with, they compose badly and without further standardization the effort will probably not outweigh the risk. If you want to know more about how to write decorators, please use StackOverflow or something similar, this is not a general JavaScript helpdesk :) 

If I might give my 2 cents on the design of your decorator, I wouldn't introduce it: `@decorator('token')` isn't statically analysable and will break whenever you name or move 'token' without any language assistance. I'd keep it explicit and just setup the observe or reaction in the constructor.

**Reply by UserAbuser228** [reply]
I would not need static analysis, and the token would always be written in Latin, which would allow me not to think about language support. I just thought that this decorator will allow to solve the DRY problem. There are quite a few places in my project where I do essentially the same thing, and replacing this routine with a single decorator call seemed like a good idea to me.

Well, I guess I was wrong in my reasoning. Thank you for your answer.
